dockerTools: Add `buildNixShellImage`

infinisil opened the pull request:
###### Motivation for this change

Edit: An updated and improved version of this is being continued in #172736

This adds `buildNixShellImage`, which can be used to create a Docker image with an environment to build a derivation, similar to a `nix-shell`, but much purer.
```nix
with import <nixpkgs> {};
dockerTools.buildNixShellImage {
  name = "hello";
  drv = hello;
}
```

```
$ nix-build hello.nix
these derivations will be built:
[...]
Finished.
/nix/store/2irs3yciz5pvi8r63bd0w16wnynx4l7b-docker-image-hello-env.tar.gz
$ docker load -i result
506c9779afde: Loading layer  10.24kB/10.24kB
[...]
96557dcba6be: Loading layer  10.24kB/10.24kB
Loaded image: hello-env:2irs3yciz5pvi8r63bd0w16wnynx4l7b
$ docker run -it hello-env:2irs3yciz5pvi8r63bd0w16wnynx4l7b
bash-4.4$ genericBuild
unpacking sources
unpacking source archive /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
[...]
patching script interpreter paths in /home/outputs/out
checking for references to /tmp/ in /home/outputs/out...
bash-4.4$ $out/bin/hello
Hello, world!
```

This PR was sponsored by [Niteo](https://niteo.co/) :sparkles: 

Ping @grahamc @nlewo @danieldk

###### Things done

- [x] Made sure it can build `pkgs.hello`, `pkgs.openssl` and `pkgs.id3v2`
- [x] Wrote docs for how to use it
- [x] Built the docs successfully

zupo commented on doc/builders/images/dockertools.xml:
```diff
@@ -475,4 +475,123 @@ buildImage {
+  <title>buildNixShellImage</title>
+
+  <para>
+    Create a Docker image that sets up the an environment to build a derivation, similar to a <command>nix-shell</command> environment.
```
```suggestion
    Create a Docker image that sets up the environment to build a derivation, similar to a <command>nix-shell</command> environment.
```

tomberek reviewed (approved):
Builds and functions as expected on NixOS. Thanks for the documentation!

danbst commented:
is it possible to mount source as a volume?
```
...
src = ./some_directory;
```
should make `$src` to point to `/source`, which can be then binded to hosts `./some_directory`. Obviously, when `src` is specified, `unpackPhase` and `patchPhase` should be run manually.

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+        passAsFile = [ "profile" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/{build,outputs}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
```
Why are you not adding `bashInteractive`, `profile`, `coreutils` and `cacert` in the `contents` instead?

Infinisil replied:
I want to only include what is absolutely needed. And that's the things added in this derivation, `/bin/sh` and `/usr/bin/env` for POSIX compatibility and the cacert thing to have SSL connectivity. All other files of these derivations aren't needed. And I put the `profile` in here too because it needs to be created in a derivation anyways, so why not use a single one for everything (instead of creating one for the symlinks and another for the profile to be put in `contents`).

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+    , drv
+    , tag ? null
+    # Extra packages available inside the image
+    , extraPackages ? []
```
What about calling this `contents`? It seems to me it behaves like the `contents` attribute of our other docker tools.

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+    let
+
+      finalDrv = drv.overrideAttrs (old: {
+        nativeBuildInputs = old.nativeBuildInputs or [] ++ extraPackages;
```
Why not just adding `extraPackages` in the contents list?

Infinisil replied:
Hm it's not the same thing: `contents` are added to the `/` path, but `extraPackages` are things that are added to `PATH` (and other env vars) later by stdenv's setup-hook. So by using `nativeBuildInputs` all kinds of dependencies should work, though now I'm debating whether that's even needed. Maybe it's needed for `man` support or so though.

nlewo replied:
IIUC, this is to allow a user to add extra tooling in its container. I think, in other images, this use case is covered by the `contents` attribute, even if it is less convenient than what you are doing (because enviroment variables are not set). In order to have a coherent `dockerTools` API, I would prefer to expose the `dockerTools.buildLayeredImage` `contents` and `env` attributes: if the user needs extra tooling, it would be still possible to add packages with contents and set some environment variables with `env`. WDYT?

roberth replied:
The meaning is different. For consistency, users will expect `contents` to be available in the root, which is not happening for `extraPackages`.
Instead, it seems like `extraPackages` should be accepted by other `dockerTools` functions, to make adding store paths easier in those rare cases that it's needed.
However, first I'd rename it to `extraStorePaths` to avoid associations with `environment.systemPackages`.

Infinisil replied:
Thinking about this again, I feel like `extraPackages` should be removed, because people can just use `.overrideAttrs` on the derivation they pass in instead.

Similarly, I don't think an `env` is needed, because this can also be achieved by setting that in the derivation in multiple ways already:
```nix
{
  VAR = "VAL";
  shellHook = ''
    export VAR=VAL
  '';
}
```

I think the idea of passing through `contents` is a good one though, I'll change it as such

nlewo reviewed (changes requested):
It would be nice to add a test in `nixos/tests/docker-tools.nix`. I think building and executing hello would be perfect ;)

tomberek commented:
still important to me

zupo commented:
Me too.

yipengsun commented:
This tool looks really nice for preserving development environment for teams that don't universally use nix! However, given that nix is introducing flakes and `nix-shell` seems to be replaced by `nix develop`, how much tweak is needed to make this work with the flake workflow?

Thanks.

3noch commented:
This would be amazing!

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+
+      # Pass --login such that /etc/profile gets loaded
+      # Use bashInteractive for tab completion support and such
+      config.Entrypoint = [ "${bashInteractive}/bin/bash" "--login" ];
```
This is different from the `bash` image, which has this entrypoint https://github.com/tianon/docker-bash/blob/master/5.1/docker-entrypoint.sh

When I try this in a normal session I get
```
$ bash --login echo foo
/run/current-system/sw/bin/echo: /run/current-system/sw/bin/echo: cannot execute binary file
```

Which seems to be a cryptic way of telling us that it's refusing to interpret an ELF executable as a script.

Passing a script does work
```
$ bash --login <(echo "echo hello") asdf
hello
```

So I think it should be like the official bash entrypoint.

The `Cmd` should probably be [`["bash"]`](https://github.com/tianon/docker-bash/blob/667f5e52f50ef8cdf3d03ab0aa86b9f90c77c726/5.1/Dockerfile#L127).

We'll need a test for this. You can add cases to `nixos/tests/docker-tools.nix`, based on a new example you can add to `pkgs/build-support/docker/examples.nix`.

Infinisil replied:
Not sure I get the motivation. `nix-shell` is bash-based, but it's not an image for the purpose of executing bash. In this case, we need some way to source stdenv and execute shellHook when it starts, which I'm now doing by passing `--login` and writing the shell init to `/etc/profile`. Currently this allows you to either do `docker run -it <image>`, then run `genericBuild`, or you can also do
```
docker run -it <image> -c genericBuild
```

And this works, which I don't think would be the case with your suggestion.

roberth replied:
I agree we seem to have different use cases in mind, but we can probably support both.

 - use the nix-shell image to build something
 - use the nix-shell image as a developer environment

For building, we should invoke the `builder` with `args` and not wrongly assume `bashInteractive`.
For interactive use, I do think we should emulate the official bash image behavior.

Can't we support both?

Infinisil replied:
I now changed it to have no entrypoint at all, however there's bash, all of coreutils and a special `buildDerivation` binary in the `PATH`. In addition, the `Cmd` is set to `[ "bash" ]`. I managed to get around the `--login` requirement by using `~/.bashrc` instead. This means by default you'll get a `nix-shell` environment, but you can also do any of
```
$ docker run -it <image> # nix-shell
$ docker run -it <image> bash # nix-shell
$ docker run -it <image> buildDerivation # build the derivation directly
$ docker run -it <image> bash -c 'buildDerivation && ls $out/bin' # builds derivation and lists the binaries it produced

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+          else toString value;
+      in lib.mapAttrsToList (name: value: "${name}=${stringValue value}") (finalDrv.drvAttrs // {
+        # These are needed for stdenv setup to work
+        NIX_STORE = "/nix/store";
```
```suggestion
        NIX_STORE = builtins.storeDir;
```

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+      # Use bashInteractive for tab completion support and such
+      config.Entrypoint = [ "${bashInteractive}/bin/bash" "--login" ];
+      config.Env = let
+        # We can't just use `toString` on all derivation attributes because that
+        # would not put path literals in the closure. So we explicitly copy
+        # those into the store here
+        stringValue = value: if builtins.typeOf value == "path" then "${value}"
+          else if builtins.isList value then toString (map stringValue value)
+          else toString value;
```
So we're basically emulating the nix sandbox env setup here. Structured attrs and passAsFile are missing, but that's probably ok for a start.

Infinisil replied:
Yeah, didn't want to bother with structuredAttrs, but I guess I can issue a warning when it's enabled

Infinisil replied:
An assertion now triggers when __structuredAttrs is set. The current code can't handle __structuredAttrs at all really.

roberth replied:
What about `passAsFile` though?

Infinisil replied:
Yeah that didn't work before, but I implemented support for it now :)

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+      # These folders should be writable
+      extraCommands = "chmod +w -R tmp home";
+
+      # Don't run this image as root because many programs don't like that
+      config.User = "1000:1000";
+      # Set the owner of the files to that user as well
+      uid = 1000;
+      gid = 1000;
```
If the user has a different uid, this breaks directory volumes for them. For this reason Arion makes the user uid available to its modules, but that's not a "luxury" we have here. Nonetheless we could make it configurable, so it's solvable without forking Nixpkgs.

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+          source $stdenv/setup
+          set +e
+          eval "$shellHook"
+          cd /home
+        '' + extraInit;
```
`extraInit` seems like a rather arbitrary term that doesn't have a well-understood purpose. Can't we remove it and rely on the user extending `shellHook` beforehand, or setting an appropriate `Cmd`?

`cd /home` breaks the `WorkingDir` config item.

Infinisil replied:
Yeah I'm removing `extraInit` in the next iteration. Can also be done with just `shellHook`.

`WorkingDir` can't be configured in the current version anyways

roberth replied:
With docker you can always configure any config attribute afterwards.

Infinisil replied:
Now using `WorkingDir` directly to specify this, instead of `cd` ing

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -886,4 +888,70 @@ rec {
+        NIX_STORE = "/nix/store";
+        NIX_BUILD_TOP = "/home/build";
+
+        HOME = "/home";
```
I'd prefer `/home/user`, to match FHS expectations.
Also `/home/build` -> `/home/user/build`, etc.

infinisil committed:
dockerTools.streamLayeredImage: Add uid/gid support

This also readds support for those arguments to buildLayeredImage, which
was removed in 560201da661a4c916d0732e3bab8a77084485659

infinisil committed:
Introduce dockerTools.buildNixShellImage

Infinisil commented:
In addition to above things, I also changed the derivation output variables to point directly at the correct `/nix/store` path, which allows `placeholder` to work correctly (tested with `bluez-alsa`). I also set a default name for the image.

Infinisil commented:
I'll update/add docs/tests/examples once I get confirmation by @nlewo and/or @roberth regarding the current implementation.

roberth commented on pkgs/build-support/docker/default.nix:
```diff
@@ -866,4 +868,77 @@ rec {
+      '';
+
+      # A mapping from output name to the nix store path where they should end up
+      outputPaths = lib.genAttrs drv.outputs (output: builtins.unsafeDiscardStringContext drv.${output}.outPath);
```
Are these paths writable in the store?
We should have a test case for this as well.

It would be nice to have symlinks to these paths in a fixed location for derived images to use.

Infinisil replied:
They are indeed writable. Perhaps I could create /home/user/outputs/* again, symlinking into the store

Infinisil replied:
Now created symlinks from `/home/user/outputs/*` into the `/nix/store` paths

tomberek commented:
Result of `nixpkgs-review pr 78967` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

3noch commented:
@Infinisil What else needs to be done for this?

roberth commented:
@3noch @Infinisil mostly just tests by the looks of it.

yipengsun commented:
For the documentation, it is using `nix-build` as an example and for `drv`, I think I can just import my `shell.nix`. However, I'm confused about how to do this for a flake-based workflow. 

Naively I'd just use the backward-compatibility tool to "expose" my `devShell` to a `shell.nix` and then build from that. Can I do something similar without any compatibility tool?

roberth commented:
@yipengsun You could move the `buildNixShellImage` call under your flake's `packages` and reference the shell derivation via `self`.
I believe there may be a function for invoking flakes but that's at least undocumented so ymmv.

3noch commented on pkgs/build-support/docker/default.nix:
```diff
@@ -866,4 +869,96 @@ rec {
+        exec ${lib.escapeShellArg (stringValue drv.drvAttrs.builder)} ${lib.escapeShellArgs (map stringValue drv.drvAttrs.args)}
+      '';
+
+    in buildLayeredImage {
```
I want to override some settings in this config. I also want to use `streamLayeredImage`. So really it makes sense to just return the attrset directly and let the user decide where the image is actually realized.

roberth replied:
You could use `(buildLayeredImage args).stream` to solve that part of the problem.
I feel the increasing number of interfaces for building docker images is starting to become a problem, so I'm not excited about having a function that returns arguments for another function. I'd like to experiment with a more powerful interface during Ocean Sprint, so I'll know more in a month.

3noch commented on pkgs/build-support/docker/default.nix:
```diff
@@ -866,4 +869,96 @@ rec {
+      inherit uid gid;
+
+      # By default just starts a shell
+      config.Cmd = [ "bash" ];
```
This should probably be easily overridable. I want to override it in my use case as I don't want to only be able to run containers interactively. Using `shellHook` to do run a task means it only works when `.bashrc` is loaded which means it must be interactive.

3noch commented:
This is important to get in.

roberth commented:
This needs tests.
It needs to test some examples and compare its behavior against regular `nix-shell`.

SuperSandro2000 commented on doc/builders/images/dockertools.xml:
```diff
@@ -496,4 +496,123 @@ buildImage {
    Creating base files like <literal>/etc/passwd</literal> or <literal>/etc/login.defs</literal> is necessary for shadow-utils to manipulate users and groups.
   </para>
  </section>
+ <section xml:id="ssec-pkgs-dockerTools-buildNixShellImage">
```
Is that file not yet converted to markdown?

Infinisil commented:
I don't have any interest in this anymore. I think this would be really nice to get in, but I'll leave it up to someone that actually wants to use it ;). Anybody feel free to use the code from this PR

infinisil commented:
Check out the new https://github.com/NixOS/nixpkgs/pull/172736, where I'm continuing work on this function
